envoy/common/conn_pool.h

```diff
@@ -44,17 +44,23 @@ class Instance {
   virtual ~Instance() = default;
 
   /**
-   * Called when a connection pool has been drained of pending streams, busy connections, and
-   * ready connections.
+   * Called when a connection pool has no pending streams, busy connections, or ready connections.
    */
-  using DrainedCb = std::function<void()>;
+  using IdleCb = std::function<void(bool is_drained)>;
+
+  enum class DrainPool { Yes, No };
+
+  /**
+   * Register a callback that gets called when the connection pool is fully idle. If `drain` is
+   * `Yes`, this also kicks off a drain. The owner of the connection pool is responsible for not
+   * creating any new streams.
+   */
+  virtual void addIdleCallback(IdleCb cb, DrainPool drain) PURE;
```

Comment:
There was a lot of commentary and discussion in the previous PR (linked in description), and it seemed the conclusion was that idle and drain callbacks were to be treated as the same. I don't have strong feelings about it either way though. @mattklein123 had the strongest opinion, so I'll let him weigh in.

Reply:
I would have to page all of this back in, but my main commentary on the previous PR was that I didn't understand the difference between drain and idle. Can't they be unified in a single callback? I can review this PR in more detail.

Reply:
I think some of the confusion is that, originally (before this PR), `addDrainCallback()` had the dual purpose of setting the callback, and starting the drain process.

It may make more sense to separate these two: when the pool is created, have the cluster manager set a callback that deletes the pool when it is called. Have this always be called if the pool goes idle. Then have another function on the pool `startDrain()` that tells the pool to try to get rid of connections, and still call the previously-registered callback when it finishes. Thoughts?

Reply:
Yes +1 that makes sense to me.

Reply:
(I will wait to review until ^ change is made if that works for you)

/wait